src/full/Agda/Syntax/Translation/ConcreteToAbstract.hs

```diff
@@ -2794,6 +2807,46 @@ instance ToAbstract LeftHandSide where
         printLocals 10 "checked dots:"
         return $ A.LHS (LHSInfo (getRange lhs) ell) lhscore
 
+-- | Expands implicit argument puns.
+
+expandPuns :: C.Pattern -> C.Pattern
+expandPuns p = case p of
+  C.AppP p1 p2       -> C.AppP (expandPuns p1)
```

Comment:
Do you mean to expand puns also in sub-patterns?  There is no test case covering this situation.
E.g.
```agda
... vcons : {n : Nat} (x : A) (xs : Vec A n) -> Vec A (suc n)

zip : Vec A k * Vec B k -> Vec (A * B) k
zip (vcons {k} x xs, vcons {k} y ys) = vcons {k} (x , y) (zip {k = k} xs ys)
```
This would become the non-sensical lhs `zip (vcons {k = k} x xs, vcons {k = k} y ys)`.

I would have expected puns only to be active at the top-level.

Reply:
I do want puns to be expanded in subpatterns:
* I think it would be strange to handle subpatterns differently.
* I have already started using this feature.

I don't think there is anything in the documentation that suggests that subpatterns are handled differently, but I can add a test case with a subpattern.

> This would become the non-sensical lhs `zip (vcons {k = k} x xs, vcons {k = k} y ys)`.

That code is fine:
```agda
{-# OPTIONS --hidden-argument-puns #-}

open import Agda.Builtin.Nat
open import Agda.Builtin.Sigma

private variable
  A B : Set
  n   : Nat

data Vec (A : Set) : Nat → Set where
  []  : Vec A zero
  _∷_ : A → Vec A n → Vec A (suc n)

zip : Σ (Vec A n) (λ _ → Vec B n) → Vec (Σ A λ _ → B) n
zip ([]           , []          ) = []
zip (_∷_ {n} x xs , _∷_ {n} y ys) =
  _∷_ {n = n} (x , y) (zip {n = n} (xs , ys))
```